base/strings/types.jl

```diff
@@ -4,32 +4,37 @@
 
 ## substrings reference original strings ##
 
+"""
+    SubString(s::AbstractString, i::Integer, j::Integer=endof(s))
+    SubString(s::AbstractString, r::UnitRange{<:Integer})
+
+Like [`getindex`](@ref), but returns a view into the parent AbstractString `s`
+with the given indices instead of making a copy.
+"""
 struct SubString{T<:AbstractString} <: AbstractString
     string::T
     offset::Int
     endof::Int
 
     function SubString{T}(s::T, i::Int, j::Int) where T<:AbstractString
-        if i > endof(s) || j<i
-            return new(s, i-1, 0)
-        else
-            if !isvalid(s,i)
-                throw(ArgumentError("invalid SubString index"))
-            end
-
-            while !isvalid(s,j) && j > i
-                j -= 1
-            end
-
-            o = i-1
-            new(s, o, max(0, j-o))
-        end
+        i > j && return new(s, i - 1, 0) # always allow i > j as it is consistent with getindex
+        isvalid(s, i) || throw(BoundsError(s, i))
+        isvalid(s, j) || throw(BoundsError(s, j))
+        new(s, i-1, j-i+1)
     end
 end
+
 SubString(s::T, i::Int, j::Int) where {T<:AbstractString} = SubString{T}(s, i, j)
-SubString(s::SubString, i::Int, j::Int) = SubString(s.string, s.offset+i, s.offset+j)
-SubString(s::AbstractString, i::Integer, j::Integer) = SubString(s, Int(i), Int(j))
-SubString(s::AbstractString, i::Integer) = SubString(s, i, endof(s))
+SubString(s::AbstractString, i::Integer, j::Integer=endof(s)) = SubString(s, Int(i), Int(j))
+SubString(s::AbstractString, r::UnitRange{<:Integer}) = SubString(s, first(r), last(r))
+
+function SubString(s::SubString, i::Int, j::Int)
+    # always allow i > j as it is consistent with getindex
+    i > j && return SubString(s.string, s.offset + i, s.offset + j)
```

Comment:
first index: I pass ` s.offset + i` and in the inner constructor it gets converted to `s.offset + i - 1` (as you have indicated).
second index: I pass `s.offset + j` and in the inner constructor it gets converted to `0` (as you have indicated).

I have to refer to `i` and `j` so that the inner constructor can know that `i>j`.
Otherwise consider the following case `s.offset = 1`, `i = -10` and `j = -20`. Under current implementation this set of parameters will go through and generate `""` as expected (to be consistent with `getindex` for `String`). If I have changed the definition to what you propose I would pass to inner constructor `SubString(s.string, -20, 1)` which would incorrectly raise `BoundsError`.